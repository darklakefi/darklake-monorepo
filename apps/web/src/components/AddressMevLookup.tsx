"use client";

import { useEffect, useRef, useState } from "react";
import Image from "next/image";
import { groupBy } from "lodash";

import { pasteFromClipboard } from "@/utils/browser";
import { isValidSolanaAddress } from "@/utils/blockchain";
import Modal from "@/components/Modal";
import { socials, SocialType } from "@/constants/links";
import { signInWithTwitter } from "@/services/supabase";
import useSupabaseSession from "@/hooks/useSupabaseSession";
import { LocalStorage } from "@/constants/storage";
import { cn } from "@/utils/common";
import Button from "@/components/Button";

const AddressMevLookup = () => {
  const [isInputVisible, setIsInputVisible] = useState(false);
  const [inputValue, setInputValue] = useState("");
  const inputRef = useRef<HTMLInputElement>(null);
  const [showResultsModal, setShowResultsModal] = useState(false);
  const [isConnectingTwitter, setIsConnectingTwitter] = useState(false);
  const supabaseSession = useSupabaseSession();

  const resetInput = () => {
    setInputValue("");
    setIsInputVisible(false);
  };

  useEffect(() => {
    if (isInputVisible) {
      inputRef.current?.focus();
    }
  }, [isInputVisible, inputRef]);

  useEffect(() => {
    if (!supabaseSession) {
      return;
    }

    const signInInitiated = localStorage.getItem(LocalStorage.SIGN_IN_INITIATED);
    if (!signInInitiated?.length) {
      return;
    }

    setShowResultsModal(true);
    localStorage.removeItem(LocalStorage.SIGN_IN_INITIATED);
  }, [supabaseSession]);

  const onPasteAddressClick = () => {
    pasteFromClipboard((pasted) => setInputValue(pasted.trim()));
    setIsInputVisible(true);
  };

  const socialsByType = groupBy(socials, "type");

  const onConnectTwitterCLick = async () => {
    if (isConnectingTwitter) return;
    setIsConnectingTwitter(true);

    await signInWithTwitter();
    localStorage.setItem(LocalStorage.LOOKUP_ADDRESS, inputValue);

    setIsConnectingTwitter(false);
  };
  const connectWithTwitterDisabled = isConnectingTwitter || !!supabaseSession;

  return (
    <div
      className={cn(
        "h-12 p-2 md:max-w-2xl w-full flex flex-row items-center justify-between border border-brand-40 bg-brand-60",
      )}
    >
      <div className="flex items-center flex-grow">
        <span className="font-secondary text-lg leading-6 text-brand-30 select-none mr-3">&gt;</span>
        {!isInputVisible && (
          <div
            className="font-secondary text-lg leading-6 w-full text-brand-30 uppercase cursor-text"
            onClick={(ref) => {
              if (ref.currentTarget === ref.target) {
                setIsInputVisible(true);
              }
            }}
          >
            Enter or{" "}
            <Button intent="secondary" className="px-3 py-1" onClick={onPasteAddressClick}>
              Paste
            </Button>{" "}
            Solana address...
          </div>
        )}
        {isInputVisible && (
          <input
            ref={inputRef}
            className={cn(
              "w-full bg-transparent min-w-0 max-w-full border-none",
              "focus:outline-none font-secondary text-lg leading-6 text-brand-20",
            )}
            type="text"
            value={inputValue}
            onBlur={() => {
              if (inputValue?.length) return;
              setIsInputVisible(false);
            }}
            onChange={(e) => setInputValue(e?.target?.value ?? "")}
          />
        )}
      </div>
      {!!inputValue?.length && (
        <button onClick={resetInput} className="hover-with-active cursor-pointer ml-2 flex items-center">
          <i className="hn hn-times-circle-solid text-brand-30 text-lg leading-5" />
        </button>
      )}
      {isValidSolanaAddress(inputValue) && (
        <button
          className={cn(
            "font-secondary text-lg leading-6 uppercase bg-brand-10 text-brand-70 px-3 py-1 hover:bg-brand-20",
            "disabled:opacity-50 focus:outline-none active:ring-1 active:bg-brand-10 active:ring-brand-10",
            "active:ring-offset-2 active:ring-offset-black ml-3 flex-shrink-0",
          )}
          onClick={() => setShowResultsModal(true)}
        >
          Reveal Losses
        </button>
      )}
      <Modal title="MEV case file: opening soon" isOpen={showResultsModal} onClose={() => setShowResultsModal(false)}>
        <div className="flex flex-col-reverse items-center lg:flex-row gap-x-20 w-full max-w-full lg:max-w-4xl">
          <div className="flex flex-row flex-wrap gap-4 w-full lg:w-[530px]">
            <div className="p-6 bg-brand-60">
              <h1 className="font-primary text-3xl leading-3xl text-brand-30 uppercase">
                <span className="text-brand-20">Connect your X account</span> so I can DM you when your MEV report is
                ready.
              </h1>
              <button
                className={cn(
                  "font-secondary text-lg leading-6 uppercase bg-brand-10 text-brand-70 px-3 py-1",
                  "hover:bg-brand-20 disabled:opacity-50 focus:outline-none active:ring-1 active:bg-brand-10",
                  "active:ring-brand-10 active:ring-offset-2 active:ring-offset-black w-full mt-6 text-center",
                )}
                onClick={onConnectTwitterCLick}
                disabled={connectWithTwitterDisabled}
              >
                {supabaseSession
                  ? `Connected as @${supabaseSession.user.user_metadata?.preferred_username}`
                  : "Connect X Account"}
              </button>
            </div>
            <div className="p-6 md:w-[calc(50%-8px)] bg-brand-60">
              <p className="font-secondary text-lg leading-6 text-brand-30 uppercase">
                <span className="text-brand-20">Join our Telegram</span> channel for real-time updates on the launch
              </p>
              <a
                href={socialsByType[SocialType.TELEGRAM][0].url}
                target="_blank"
                rel="noreferrer"
                title="Join Darklake Telegram"
                className={cn(
                  "font-secondary text-lg leading-6 uppercase underline bg-brand-50 text-brand-20 px-3 py-1",
                  "hover:text-brand-10 disabled:opacity-50 focus:outline-none active:ring-1 active:bg-brand-50",
                  "active:ring-brand-20 active:ring-offset-2 active:ring-offset-black; block mt-[24px] text-center",
                )}
              >
                Join Telegram
              </a>
            </div>
            <div className="p-6 md:w-[calc(50%-8px)] bg-brand-60">
              <p className="font-secondary text-lg leading-6 text-brand-30 uppercase">
                <span className="text-brand-20">Follow @darklakefi on X</span> for daily updates on our investigation
              </p>
              <a
                href={socialsByType[SocialType.TWITTER][0].url}
                target="_blank"
                rel="noreferrer"
                title="Follow Darklake on X"
                className={cn(
                  "bg-brand-50 text-brand-20 px-3 py-1 block mt-6",
                  "font-secondary text-lg leading-6 uppercase underline text-center",
                  "hover:text-brand-10 disabled:opacity-50 focus:outline-none",
                  "active:ring-1 active:bg-brand-50 active:ring-brand-20 active:ring-offset-2 active:ring-offset-black",
                )}
              >
                Follow on X
              </a>
            </div>
          </div>
          <Image
            src="/images/image-waddles-1.png"
            alt="darklake waddles 1"
            width={350}
            height={430}
            className="relative lg:bottom-[-40px] hidden lg:block"
          />
        </div>
      </Modal>
    </div>
  );
};

export default AddressMevLookup;
