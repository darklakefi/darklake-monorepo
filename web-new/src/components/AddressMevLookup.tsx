"use client";

import { useEffect, useRef, useState } from "react";
import Image from "next/image";
import { groupBy } from "lodash";

import { pasteFromClipboard } from "@/utils/browser";
import { isValidSolanaAddress } from "@/utils/blockchain";
import Modal from "@/components/Modal";
import { socials, SocialType } from "@/constants/links";

import iconCloseCircle from "../../public/images/icon-close-circle.png";
import imageWaddles1 from "../../public/images/image-waddles-1.png";

const AddressMevLookup = () => {
  const [isInputVisible, setIsInputVisible] = useState(false);
  const [inputValue, setInputValue] = useState("");
  const inputRef = useRef<HTMLInputElement>(null);
  const [showResultsModal, setShowResultsModal] = useState(false);

  const resetInput = () => {
    setInputValue("");
    setIsInputVisible(false);
  };

  useEffect(() => {
    if (isInputVisible) {
      inputRef.current?.focus();
    }
  }, [isInputVisible, inputRef]);

  const onPasteAddressClick = () => {
    pasteFromClipboard((pasted) => setInputValue(pasted));
  };

  const socialsByType = groupBy(socials, "type");

  return (
    <div className="h-[50px] p-[8px] md:w-[628px] w-full flex flex-row items-center justify-between border border-brand-40 bg-brand-60">
      <div className="flex items-center flex-grow">
        <span className="text-body-2 text-brand-30 select-none mr-[13px]">&gt;</span>
        {!isInputVisible && (
          <div
            className="text-body-2 w-full text-brand-30 uppercase cursor-text"
            onClick={(ref) => {
              if (ref.currentTarget === ref.target) {
                setIsInputVisible(true);
              }
            }}
          >
            Enter or{" "}
            <button className="button-secondary" onClick={onPasteAddressClick}>
              Paste
            </button>{" "}
            Solana address...
          </div>
        )}
        {isInputVisible && (
          <input
            ref={inputRef}
            className="w-full bg-transparent min-w-0 max-w-full border-none focus:outline-none text-body-2 text-brand-20"
            type="text"
            value={inputValue}
            onBlur={() => {
              if (inputValue?.length) return;
              setIsInputVisible(false);
            }}
            onChange={(e) => setInputValue(e?.target?.value ?? "")}
          />
        )}
      </div>
      {!!inputValue?.length && (
        <Image
          src={iconCloseCircle}
          onClick={resetInput}
          className="active-hover-opacity cursor-pointer ml-[8px]"
          alt="close"
        />
      )}
      {isValidSolanaAddress(inputValue) && (
        <button className="button-primary-light ml-[12px] flex-shrink-0" onClick={() => setShowResultsModal(true)}>
          Reveal Losses
        </button>
      )}
      <Modal title="MEV case file: opening soon" isOpen={showResultsModal} onClose={() => setShowResultsModal(false)}>
        <div className="flex flex-col-reverse items-center lg:flex-row gap-x-[80px] w-full lg:w-[960px]">
          <div className="flex flex-row flex-wrap gap-[16px] w-full lg:w-[530px]">
            <div className="p-[24px] bg-brand-60">
              <p className="text-heading-1 text-brand-30 uppercase">
                <span className="text-brand-20">Connect your X account</span> so I can DM you when your MEV report is
                ready.
              </p>
              <button className="button-primary-light w-full mt-[24px] text-center">Connect X Account</button>
            </div>
            <div className="p-[24px] md:w-[calc(50%-8px)] bg-brand-60">
              <p className="text-body-2 text-brand-30 uppercase">
                <span className="text-brand-20">Join our Telegram</span> channel for real-time updates on the launch
              </p>
              <a
                href={socialsByType[SocialType.TELEGRAM][0].url}
                target="_blank"
                rel="noreferrer"
                title="Join Darklake Telegram"
                className="button-secondary block mt-[24px] text-center"
              >
                Join Telegram
              </a>
            </div>
            <div className="p-[24px] md:w-[calc(50%-8px)] bg-brand-60">
              <p className="text-body-2 text-brand-30 uppercase">
                <span className="text-brand-20">Follow @darklakefi on X</span> for daily updates on our investigation
              </p>
              <a
                href={socialsByType[SocialType.TWITTER][0].url}
                target="_blank"
                rel="noreferrer"
                title="Follow Darklake on X"
                className="button-secondary block mt-[24px] text-center"
              >
                Follow on X
              </a>
            </div>
          </div>
          <Image src={imageWaddles1} alt="darklake waddles 1" className="relative lg:bottom-[-40px] hidden lg:block" />
        </div>
      </Modal>
    </div>
  );
};

export default AddressMevLookup;
