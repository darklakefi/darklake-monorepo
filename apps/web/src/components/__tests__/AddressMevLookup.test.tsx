import { act, render, RenderResult } from "@testing-library/react";
import AddressMevLookup from "@/components/AddressMevLookup";

describe("<AddressMevLookup />", () => {
  it("renders successfully", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(<AddressMevLookup />));
    });

    expect(baseElement).toMatchSnapshot();
  });
});
