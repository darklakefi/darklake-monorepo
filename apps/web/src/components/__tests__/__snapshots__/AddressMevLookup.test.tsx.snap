// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressMevLookup /> renders successfully 1`] = `
<body>
  <div>
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="h-12 p-2 md:max-w-2xl w-full bg-brand-60 border border-brand-40 hover:border-brand-30 flex flex-row items-center"
      >
        <div
          class="flex items-center w-full"
        >
          <span
            class="font-secondary text-lg leading-6 tracking-normal text-brand-30 select-none mr-3"
          >
            &gt;
          </span>
          <div
            class="font-secondary text-lg max-md:text-xs leading-6 w-full text-brand-30 uppercase cursor-text"
          >
            Enter or
             
            <button
              class="uppercase font-secondary button text-lg leading-6 tracking-normal underline bg-brand-50 text-brand-20 hover:bg-brand-50 focus:outline-none active:ring-[1px] active:bg-brand-50 active:ring-brand-20 active:ring-offset-[3px] active:ring-offset-black disabled:active:bg-brand-50 px-3 py-1 max-md:text-xs"
            >
              Paste
            </button>
             
            Solana address...
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
